### PR TITLE
Add showing of model connectivity

### PR DESCRIPTION
**Pull request by graeme-winter**
Shows how models are related to experiments e.g. 

```
Experiment / Models

Detector:
              0  1
Experiment 0  X  /
Experiment 1  X  /
Experiment 2  X  /
Experiment 3  /  X

Crystal:
              0  1  2  3
Experiment 0  X  /  /  /
Experiment 1  /  X  /  /
Experiment 2  /  /  X  /
Experiment 3  /  /  /  X

Beam:
              0
Experiment 0  X
Experiment 1  X
Experiment 2  X
Experiment 3  X
```

Seemed useful to me... make an option?

### CODE REVIEW

**Comment by graeme-winter** [comment]
N.B. allows clearer view of what options do e.g. with and without `split_experiments` in `dials.index`

```
Experiment / Models

Detector:
              0  1  2  3
Experiment 0  X  /  /  /
Experiment 1  /  X  /  /
Experiment 2  /  /  X  /
Experiment 3  /  /  /  X

Crystal:
              0  1  2  3
Experiment 0  X  /  /  /
Experiment 1  /  X  /  /
Experiment 2  /  /  X  /
Experiment 3  /  /  /  X

Beam:
              0  1  2  3
Experiment 0  X  /  /  /
Experiment 1  /  X  /  /
Experiment 2  /  /  X  /
Experiment 3  /  /  /  X
```

vs.

```
Experiment / Models

Detector:
              0  1
Experiment 0  X  /
Experiment 1  X  /
Experiment 2  X  /
Experiment 3  /  X

Crystal:
              0
Experiment 0  X
Experiment 1  X
Experiment 2  X
Experiment 3  X

Beam:
              0
Experiment 0  X
Experiment 1  X
Experiment 2  X
Experiment 3  X
```

**Comment by graeme-winter** [comment]
Does not show anything if `len(experiments) == 1` as obviously the connectivity is trivial. 

Should we have a limit for e.g. > 20 experiments or something? Can see this being useful for indexing > 1 lattice across > 1 sweep though. 

Also _welcome bikeshedding_ on `X` vs. `/` etc. - what are better symbols? - should we have `---` and `|` to make ersatz ASCII art table? If you feel yes, please feel free to add 😉

File: command_line/show.py [hunk 1]
```diff
@@ -299,6 +301,58 @@ def show_experiments(experiments, show_scan_varying=False, show_image_statistics
...
+    detectors = experiments.detectors()
+    beams = experiments.beams()
+    crystals = experiments.crystals()
+
+    text.append("")
+    text.append("Experiment / Models")
+    text.append("")
+    text.append("Detector:")
+    text.append("              " + "  ".join([str(j) for j in range(len(detectors))]))
+    for j, e in enumerate(experiments):
```
[the remark this answers is not in the document]
Comment (graeme-winter) [reply command_line/show.py]: Don't know about `tabulate` however sounds like _a good thing_ so 100% on board.

File: command_line/show.py [hunk 2]
```diff
@@ -299,6 +301,58 @@ def show_experiments(experiments, show_scan_varying=False, show_image_statistics
     return "\n".join(text)
 
 
+def show_connectivity(experiments):
```
[the remark this answers is not in the document]
Comment (graeme-winter) [reply command_line/show.py]: `experiment_has_model` is exactly what it says on the tin

File: command_line/show.py [hunk 3]
```diff
@@ -209,6 +209,8 @@ def run(args):
             )
         )
 
+        print(show_connectivity(experiments))
```
[the remark this answers is not in the document]
Comment (graeme-winter) [reply command_line/show.py]: +1

**Comment by graeme-winter** [comment]
Test case data added to https://github.com/dials/dials/pull/995

File: command_line/show.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Anthchirp) [reply command_line/show.py]: if you end up using dependencies in dials please add them explicitly to dials (ie. `libtbx_config`) even if they are already available indirectly

**Comment by graeme-winter** [comment]
Added `show_experiment_has_model` bool param.

Re: tabulate; no idea; not learning but would very much welcome someone else dropping this in there modulo comments made by @Anthchirp re: dependencies

**Comment by dagewa** [comment]
`tabulate` looks interesting, but we have lots of uses of `simple_table` in the codebase at the moment. If we are going to include `tabulate` I suggest it is done outside this PR while converting the existing `simple_table`s for consistency.

File: command_line/show.py [hunk 4]
```diff
@@ -26,6 +27,9 @@
 show_scan_varying = False
   .type = bool
   .help = "Whether or not to show the crystal at each scan point."
+show_experiment_has_model = False
```
Comment (rjgildea) [inline command_line/show.py]: Perhaps this would be a better parameter (and relevant function) name?
```suggestion
show_model_connectivity = False
```

**Comment by graeme-winter** [comment]
Am happy for anyone to take this one forward to close off :)

File: test/command_line/test_show.py [hunk 5]
```diff
@@ -365,3 +368,46 @@ def test_dials_show_on_scaled_data(dials_data):
...
+Experiment 1  X
+Experiment 2  X
+Experiment 3  X"""
+    )
+
+
+def test_dials_show_model_connectivity(dials_data):
+    """Test that dials.show experiments_has_model option."""
+    location = dials_data("l_cysteine_dials_output")
+    expt = location.join("indexed.expt").strpath
```
Comment (Anthchirp) [inline test/command_line/test_show.py]: don't need to `.strpath` for procrunner

File: test/command_line/test_show.py [hunk 6]
```diff
@@ -365,3 +368,46 @@ def test_dials_show_on_scaled_data(dials_data):
...
+    result = procrunner.run(["dials.show", expt, "show_model_connectivity=True"])
+    assert not result.returncode and not result.stderr
+    assert "Experiment / Models" in result.stdout
```
Comment (Anthchirp) [inline test/command_line/test_show.py]:
```suggestion
    assert b"Experiment / Models" in result.stdout
```

File: command_line/show.py [hunk 7]
```diff
@@ -299,6 +307,35 @@ def show_experiments(experiments, show_scan_varying=False, show_image_statistics
...
+        text.append("%s:" % model.capitalize())
+        models = getattr(experiments, "%ss" % model)()
+        rows = [[""] + [str(j) for j in range(len(models))]]
+        for j, e in enumerate(experiments):
+            row = ["Experiment %d" % j]
+            for m in models:
+                if getattr(e, model) is m:
+                    row.append("X")
+                else:
+                    row.append("/")
```
Comment (Anthchirp) [inline command_line/show.py]:
```suggestion
                    row.append(".")
```

my :bikeshed: contribution.

**Comment by rjgildea** [comment]
@graeme-winter I think `dials.find_shared_models` is relevant to this (although it doesn't appear to be tested) - perhaps given that similar functionality exists elsewhere then this PR is not needed?

```
$ dials.find_shared_models $(dials.data get -q l_cysteine_dials_output)/indexed.expt
DIALS (2018) Acta Cryst. D74, 85-97. https://doi.org/10.1107/S2059798317017235
DIALS 2.dev.1067-g89d664ba9
The following parameters have been modified:

input {
  experiments = /Users/rjgildea/software/cctbx/build/dials_data/l_cysteine_dials_output/indexed.expt
}

Number of sequences = 4
Sorting sequences based on timestamp
4 datasets collected on 2016-03-05
=====================================================  ====  ======  ==========  ============  ==========  ========
Sequence                                                 ID    Beam    Detector    Goniometer  Date        Time
=====================================================  ====  ======  ==========  ============  ==========  ========
/Users/graeme/data/i19-1-l-cys/l-cyst_04_#####.cbf.gz     0       0           0             0  2016-03-05  08:58:02
/Users/graeme/data/i19-1-l-cys/l-cyst_01_#####.cbf.gz     1       0           1             1  2016-03-05  09:01:45
/Users/graeme/data/i19-1-l-cys/l-cyst_02_#####.cbf.gz     2       0           1             2  2016-03-05  09:04:57
/Users/graeme/data/i19-1-l-cys/l-cyst_03_#####.cbf.gz     3       0           1             3  2016-03-05  09:08:09
=====================================================  ====  ======  ==========  ============  ==========  ========
```

**Comment by dagewa** [comment]
I think `dials.show` is a good home for this, but I prefer `show_shared_models` as a parameter name over `show_model_connectivity` Perhaps the behaviour of `dials.find_shared_models` and this PR can be merged, and `dials.find_shared_models` then retired?

**Comment by graeme-winter** [comment]
@dagewa `show_shared_models` - fine with this - I'll try to get this cleaned up though looks like we've acquired some conflicts along the way.

**Comment by rjgildea** [comment]
@graeme-winter any comment on https://github.com/dials/dials/pull/996#issuecomment-551110765? It seems that this essentially does the same thing as this PR, albeit in a different form. If this PR _is_ merged, then I suggest that `dials.find_shared_models` should be removed as it is then duplicate (and untested) functionality.

**Comment by Anthchirp** [comment]
I will spin that out as a separate issue to follow up at some point in the indefinite future

**Pull request merged**